### PR TITLE
add missing TokenType entries to TOKEN_TYPE_NAMES

### PR DESCRIPTION
**Pull request by quasilyte**
To keep these in sync in the future, assert has been added.

Signed-off-by: Iskander Sharipov <quasilyte@gmail.com>

### CODE REVIEW

File: compiler/debug.cpp [hunk 1]
```diff
@@ -161,14 +172,20 @@ void fillTokenTypeNames() {
...
   TOKEN_TYPE_NAMES[tok_private] = "tok_private";
   TOKEN_TYPE_NAMES[tok_protected] = "tok_protected";
+  TOKEN_TYPE_NAMES[tok_phpdoc] = "tok_phpdoc";
   TOKEN_TYPE_NAMES[tok_clone] = "tok_clone";
   TOKEN_TYPE_NAMES[tok_instanceof] = "tok_instanceof";
   TOKEN_TYPE_NAMES[tok_end] = "tok_end";
+
+  // this assert will fail whether new TokenType entry is added;
+  // how to fix: add TOKEN_TYPE_NAMES[$newtok] = "$newtok" to the block above
+  kphp_assert((TOKEN_TYPE_NAMES.size()-1) == tok_end);
```
[the remark this answers is not in the document]
Comment (quasilyte) [reply compiler/debug.cpp:188]: Good point.

**Review by AlexK0 — DISMISSED** [review]

**Review by eugene536 — CHANGES_REQUESTED** [review]
You don't need such comprehensive logic.

1) std::vector is not necessary here, just left `std::array` which you get from `vk::to_array` and mark it  with `constexpr` keyword for the future benefits

2) sort elements is OTT here, it's better to see enum elements sorted, therefore it's better just putting `kphp_assert(std::is_sorted(...))` and transform it to `static_assert` from c++20 (#if __cplusplus > 201703)

3) checking `TokenType` on <0 and >size is awkward IMHO because it must be checked by a type system, making an assert instead of `if` would be better 

[Here](https://pastebin.com/3zL7N45B) you can find a diff how I see it.

image

**Comment by quasilyte** [comment]
I remember the times when this PR was about adding missing entries to the token strings map. :sweat_smile:

**Comment by quasilyte** [comment]
> 3. checking TokenType on <0 and >size is awkward IMHO because it must be checked by a type system, making an assert instead of if would be better

I would keep the assert just in case this function is called with something that was casted to the token type from something like `int`.

> 2. std::vector is not necessary here, just left std::array which you get from vk::to_array and mark it with constexpr keyword for the future benefits

Well, it was an array before I thought that we might replace an array of pairs with array (well, it could be a vector) of just strings that is indexed by enum value directly. :)

**Comment by quasilyte** [comment]
If we put kphp_assert for `is_sorted` (until it can be a static_assert) I believe it'll have to do a full array traversal every time a function is called.

**Comment by quasilyte** [comment]
Removed the lambda, added a static bool variable to keep the sorting result.

**Review by AlexK0 — APPROVED** [review]

**Review by eugene536 — APPROVED** [review]

**Pull request merged**